crowbar_framework/lib/crowbar/deployment_queue.rb

```diff
@@ -0,0 +1,394 @@
+
+        depprop && !depprop_queued && depprop_deployed
+      end
+    end
+
+
+    def file_lock
+      FileLock
+    end
+
+    # Removes the proposal reference from the queue, updates the proposal as not queued
+    # and drops the 'pending roles' from the affected nodes.
+    def dequeue_proposal_no_lock(bc, inst)
+      logger.debug("dequeue_proposal_no_lock: enter #{inst} #{bc}")
+      begin
+        # Find the proposal to delete, get its elements (nodes)
+        item = proposal_queue.proposals.find { |i| i["barclamp"] == bc && i["inst"] == inst }
+
+        if item
+          elements = item["elements"]
+          proposal_queue.delete(item)
+
+          # Remove the pending roles for the current proposal from the node records.
+          remove_pending_elements(bc, inst, elements) if elements
+        end
+
+        # Mark the proposal as not in the queue
+        prop = Proposal.where(barclamp: bc, name: inst).first
+        unless prop.nil?
+          prop["deployment"][bc]["crowbar-queued"] = false
+          prop.save
+        end
+      rescue StandardError => e
+        logger.error("Error dequeuing proposal for #{bc}:#{inst}: #{e.message} #{e.backtrace.join("\n")}")
+        logger.debug("dequeue proposal_no_lock: exit #{inst} #{bc}: error")
+        return false
+      end
+      logger.debug("dequeue proposal_no_lock: exit #{inst} #{bc}")
+      true
+    end
+
+    # Each node keeps a list of roles (belonging to the current proposal) that
+    # are to be applied to it under crowbar.pending.barclamp-name hash.
+    # When we finish deploying and also when we dequeue the proposal, the list
+    # should be emptied.  FIXME: looks like bc-inst: value should be a list, not
+    # a hash?
+    def remove_pending_elements(bc, inst, elements)
+      nodes_map = elements_to_nodes_to_roles_map(elements)
+
+      # Remove the entries from the nodes.
+      f = file_lock.acquire("BA-LOCK", logger: logger)
+      begin
+        nodes_map.each do |node_name, data|
+          node = NodeObject.find_node_by_name(node_name)
+          next if node.nil?
+          unless node.crowbar["crowbar"]["pending"].nil? or node.crowbar["crowbar"]["pending"]["#{bc}-#{inst}"].nil?
+            node.crowbar["crowbar"]["pending"]["#{bc}-#{inst}"] = {}
+            node.save
+          end
+        end
+      ensure
+        file_lock.release(f)
+      end
+    end
+
+    # Create map with nodes and their element list
+    # Transform ( {role => [nodes], role1 => [nodes]} hash to { node => [roles], node1 => [roles]},
+    # accounting for clusters
+    def elements_to_nodes_to_roles_map(elements, element_order = [])
+      nodes_map = {}
+      active_elements = element_order.flatten
+
+      elements.each do |role_name, nodes|
+        next unless active_elements.include?(role_name)
+
+        # Expand clusters to individual nodes
+        nodes, failures = expand_nodes_for_all(nodes)
```

Comment:
Hrm, don't we need to have `ServiceObject.expand_nodes_for_all` or something like this?